### PR TITLE
WIP: Support cabal

### PR DESCRIPTION
**Pull request by gleber**
Sharing mostly so that you can make use of the refactorings of scripts.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,6 +33,9 @@ binary in the PATH.
 unzip -j protoc-*-linux-x86_64.zip bin/protoc -d /usr/local/bin/
 ```
 
+> If you use have Nix installed and you use `stack --nix`, you do not need to do
```
Comment (robinp) [inline README.md]: Please rebase your branch on google/master and repush (maybe with --force), as this is already on master.

File: build-cabal.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
+
+# Script for building Kythe index of Haskell packages.
+# Usage: bash build-stack.sh /tmp/logs package...
```
Comment (robinp) [inline build-cabal.sh]: s/stack/cabal

File: build-stack.sh [hunk 3]
```diff
@@ -21,27 +21,8 @@ export REALGHC=$(stack path --compiler-exe)
 # ghc_kythe_wrapper (from local-install-root) on the PATH. Note that stack
 # wrapper ghc script replaces the system ghc.
 PATH=$PWD/wrappers/stack:$(stack path --compiler-bin):$PATH:$(stack path --local-install-root)/bin \
-  stack --system-ghc build --force-dirty ${@:2}
+  stack --system-ghc build --no-nix --force-dirty ${@:2}
```
Comment (robinp) [inline build-stack.sh]: Add comment why `--no-nix` is beneficial here.

File: scripts/common.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    do
+        echo " * ${e}"
+        cat ${e} | /opt/kythe/tools/write_entries --graphstore ${INDEXER_OUTPUT_DIR}/gs
+    done
+    echo "== Converting to serving tables."
+    /opt/kythe/tools/write_tables \
+        --graphstore ${INDEXER_OUTPUT_DIR}/gs \
+        --out ${INDEXER_OUTPUT_DIR}/tbl \
+        --compress_shards
+    echo "== Starting HTTP server."
```
Comment (robinp) [inline scripts/common.sh]: could split to 2 funs: build_index for write_entries+write_tables, serve_index for the http serving

File: wrappers/cabal/ghc [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+# GHC wrapper for indexing Haskell packages via Stack.
+# Note that variables INDEXER_OUTPUT_DIR and REALGHC are set outside this script.
+
+PKG=$(basename "$(pwd)")
+
+mkdir -p "${INDEXER_OUTPUT_DIR}"
```
Comment (robinp) [inline wrappers/cabal/ghc]: please copy the dir checks that got added to the other wrapper in the mean time

**Comment by robinp** [comment]
I'll get back to this, sorry for the delay. I have some experimental indexer that uses the GHC frontend plugin with the `-w` option of certain cabal commands.

**Comment by mpickering** [comment]
@robinp Is the code for the plugin version available anywhere?

**Comment by robinp** [comment]
@mpickering not really, didn't have time to look into this recently. The gist at https://gist.github.com/robinp/5b2494c2216eeb1c4d7bf090fa59bc96 has the skeleton of the GHC wrapper that will instrument using the plugin (but this wrapper can as well be a shell script, no heavy Haskell logic here).

I was following the blogpost http://blog.ezyang.com/2017/02/how-to-integrate-ghc-api-programs-with-cabal. The main benefit I would expect from using the plugin is the automatic parsing of the DynFlags. Actually parsing (most of the) DynFlags is easy, there's a lib call for that. But it's less easy to set the proper `hscTarget` and `ghcLink` options (though I have a rough graps on that already, to be implemented in pluginless mode yet). If the plugin would set those too properly, that would be great.

In case you would like to add this support, feel free to go forward. I would be thankful.

**Comment by robinp** [comment]
Actually based on http://stuff.codereview.me/#ghc/ghc/Main.hs?corpus=ghc-8.2.1-rc2&signature&line=163 (see catchall case, which covers `DoFrontend` ctor) it sets the compile+link flags. Well, not dark magic, maybe I overcomplicated.

**Comment by robinp** [comment]
@mpickering: just to clarify, my impression is you are working on this now. Do you also chase the Frontend Plugin route, or focusing on wrapping cabal properly?

If you do the Frontend Plugin too, I would be interested in some design / early sneak peak, since I imagine changes the way the parts are plumbed together quite a bit.

For example I imagine ghc-kythe-wrapper's frontend alternative would call the indexer (in the Ghc monad), in contrast to the current wrapper where the indexer calls `GhcApiSupport`.

**Comment by mpickering** [comment]
I do intend to work on this but I got sidetracked trying to package kythe for nix which has turned out to be quite a struggle. 

I haven't looked at how things are currently structured yet but my motivation is that I want to run the indexer as a fronted plugin rather than using a wrapper so that cabal doesn't get confused about which versions of ghc-pkg to use for example.

**Comment by robinp** [comment]
Sounds great! FYI for general Kythe building questions, the devs at the https://groups.google.com/forum/#!forum/kythe mailing list are quite responsive.

**Comment by robinp** [comment]
Closing in favor of recent commit adding a build-cabal.sh, supporting the new- (or v2-) builds.

**Pull request closed without merging**